d/main.d

```diff
@@ -0,0 +1,142 @@
+import std.stdio;
+import std.random;
+
+private struct Node
+{
+    int x = 0;
+    int y = 0;
+
+    this(int x)
+    {
+        this.x = x;
+        y = uniform!int();
+    }
+
+    Node* left = null;
+    Node* right = null;
+}
+
+struct Tree
+{
+    bool hasValue(int x)
+    {
+        Node* lower, equal, greater;
+        split(mRoot, lower, equal, greater, x);
+        bool res = equal != null;
+        mRoot = merge(lower, equal, greater);
+        return res;
+    }
+
+    void insert(int x)
+    {
+        Node* lower, equal, greater;
+        split(mRoot, lower, equal, greater, x);
+        if(!equal)
+            equal = new Node(x);
+
+        mRoot = merge(lower, equal, greater);
+    }
+
+    void erase(int x)
+    {
+        Node* lower, equal, greater;
+        split(mRoot, lower, equal, greater, x);
+        mRoot = merge(lower, greater);
+        clear(equal);
+    }
+
+private:
+    
+
+    Node* mRoot = null;
+};
+
+void clear(Node* node)
+{
+    if(node == null)
+        return;
+
+    clear(node.left);
+    clear(node.right);
+}
```

Comment:
This clear function doesn't do anything here, so there is no reason to leave it. In C++ it performed removing.